contracts/governor/src/contract.rs

```diff
@@ -1,4 +1,209 @@
+        if proposal.is_none() {
+            panic_with_error!(&e, GovernorError::NonExistentProposalError);
+        }
+
+        let proposal = proposal.unwrap_optimized();
+        if e.ledger().timestamp() < proposal.vote_end {
+            panic_with_error!(&e, GovernorError::VotePeriodNotFinishedError)
+        }
+
+        let settings = storage::get_settings(&e);
+        let total_vote_supply =
+            VotesClient::new(&e, &storage::get_voter_token_address(&e)).total_supply();
+        let mut quorum_votes: i128 = 0;
+        let proposal_results = storage::get_proposal_vote_count(&e, &proposal_id);
+
+        if settings.counting_type & 0x1 == 1 {
+            quorum_votes += proposal_results.votes_abstained;
+        }
+        if settings.counting_type >> 1 & 0x1 == 1 {
+            quorum_votes += proposal_results.votes_against;
+        }
+        if settings.counting_type >> 2 & 0x1 == 1 {
+            quorum_votes += proposal_results.votes_for;
+        }
+        let votes_for_percent = proposal_results.votes_for * 100
+            / (proposal_results.votes_against + proposal_results.votes_for);
+
+        if (quorum_votes * 100 / total_vote_supply) as u32 >= settings.quorum
+            && votes_for_percent as u32 >= settings.vote_threshold
+        {
+            storage::set_proposal_status(&e, &proposal_id, &ProposalStatus::Queued);
+        } else {
+            storage::set_proposal_status(&e, &proposal_id, &ProposalStatus::Defeated);
+        }
+    }
+
+    fn execute(e: Env, proposal_id: u32) {
+        let proposal = storage::get_proposal(&e, &proposal_id);
+        if proposal.is_none() {
+            panic_with_error!(&e, GovernorError::NonExistentProposalError);
+        }
+
+        let proposal = proposal.unwrap();
+        let status = storage::get_proposal_status(&e, &proposal_id);
+        let settings = storage::get_settings(&e);
+        if status != ProposalStatus::Queued {
+            panic_with_error!(&e, GovernorError::ProposalNotQueuedError);
+        }
+        if e.ledger().timestamp() < proposal.vote_end + settings.timelock {
+            panic_with_error!(&e, GovernorError::TimelockNotMetError);
+        }
+
+        let mut pre_auth_vec: Vec<InvokerContractAuthEntry> = vec![&e];
+        for call_data in proposal.sub_calldata {
+            let pre_auth_entry = InvokerContractAuthEntry::Contract(SubContractInvocation {
+                context: ContractContext {
+                    contract: call_data.contract_id,
+                    fn_name: call_data.function,
+                    args: call_data.args,
+                },
+                sub_invocations: vec![&e],
+            });
+            pre_auth_vec.push_back(pre_auth_entry);
+        }
+        e.authorize_as_current_contract(pre_auth_vec);
+        let _: () = e.invoke_contract(
+            &proposal.calldata.contract_id,
+            &proposal.calldata.function,
+            proposal.calldata.args,
+        );
```

Comment:
what type is _?

Feels like it might make sense to return this back to the caller?

Reply:
If we want to return the value we need to store the return type of the contract call in the call data. This notation is weird the other option would be 
`e.invoke_contract::<()>(
     &proposal.calldata.contract_id,
     &proposal.calldata.function,
     proposal.calldata.args,)`